tools/clang/unittests/HLSLExec/CoopVec.h

```diff
@@ -354,6 +359,520 @@ GetMatrixSrcDataType(D3D12_LINEAR_ALGEBRA_DATATYPE MatrixInterpretation) {
+                       const TestVector &Bias, bool HasBias,
+                       D3D12_LINEAR_ALGEBRA_DATATYPE MatrixInterpretation,
+                       D3D12_LINEAR_ALGEBRA_DATATYPE InputType) {
+    // The CPU reference matrix is FP32 for all FP interpretations.
+    bool IsMatrixFP32 = false;
+    switch (MatrixInterpretation) {
+    case D3D12_LINEAR_ALGEBRA_DATATYPE_FLOAT16:
+    case D3D12_LINEAR_ALGEBRA_DATATYPE_FLOAT_E4M3:
+    case D3D12_LINEAR_ALGEBRA_DATATYPE_FLOAT_E5M2:
+      IsMatrixFP32 = true;
+      break;
+    default:
+      break;
+    }
+
+    TestVector ResultVec(InputVector.getNumVectors(), Matrix.getNumVectors(),
+                         sizeof(float));
+
+    if (IsMatrixFP32) {
+      for (size_t VecIdx = 0; VecIdx < InputVector.getNumVectors(); ++VecIdx) {
+        const DirectX::PackedVector::HALF *InputBiasFP16 =
+            Bias.getVector<DirectX::PackedVector::HALF>(0);
+        for (size_t OutputIdx = 0; OutputIdx < Matrix.getNumVectors();
+             ++OutputIdx) {
+          float Acc = 0;
+
+          for (size_t InputIdx = 0; InputIdx < Matrix.getVectorSize();
+               ++InputIdx) {
+            float InputElem;
+            if (InputType == D3D12_LINEAR_ALGEBRA_DATATYPE_FLOAT32)
+              InputElem = InputVector.getVector<float>(VecIdx)[InputIdx];
+            else
+              InputElem = ConvertFloat16ToFloat32(
+                  InputVector.getVector<DirectX::PackedVector::HALF>(
+                      VecIdx)[InputIdx]);
+
+            float const MatrixElem =
+                Matrix.getVector<float>(OutputIdx)[InputIdx];
+            Acc += InputElem * MatrixElem;
+          }
+
+          if (HasBias)
+            Acc += ConvertFloat16ToFloat32(InputBiasFP16[OutputIdx]);
+
+          float Result = Acc;
+          ResultVec.getVector<float>(VecIdx)[OutputIdx] = Result;
+        }
+      }
+    } else if (MatrixInterpretation == D3D12_LINEAR_ALGEBRA_DATATYPE_SINT8) {
+      for (size_t VecIdx = 0; VecIdx < InputVector.getNumVectors(); ++VecIdx) {
+        const int32_t *InputBiasI32 = Bias.getVector<int32_t>(0);
+        for (size_t OutputIdx = 0; OutputIdx < Matrix.getNumVectors();
+             ++OutputIdx) {
+          int Acc = 0;
+
+          for (size_t InputIdx = 0; InputIdx < Matrix.getVectorSize();
+               ++InputIdx) {
+            int InputElem;
+            if (InputType == D3D12_LINEAR_ALGEBRA_DATATYPE_FLOAT32)
+              InputElem = static_cast<int>(
+                  InputVector.getVector<float>(VecIdx)[InputIdx]);
+            else
+              InputElem = InputVector.getVector<int8_t>(VecIdx)[InputIdx];
+
+            int const MatrixElem =
+                Matrix.getVector<int8_t>(OutputIdx)[InputIdx];
+            Acc += InputElem * MatrixElem;
+          }
+
+          if (HasBias)
+            Acc += InputBiasI32[OutputIdx];
+
+          float Result = float(Acc);
+          ResultVec.getVector<float>(VecIdx)[OutputIdx] = Result;
```

Comment:
>ResultVec.getVector<float>(VecIdx)[OutputIdx] = Result; [](http://example.com/codeflow?start=10&length=55)

ResultVec.getVector<float>(VecIdx)[OutputIdx] = Acc;

Reply:
Ah, It's an int in this case. Preview block its already a float.

ResultVec.getVector<float>(VecIdx)[OutputIdx] = static_cast<float>(Acc);